### ISSUE TITLE
Erroneous chunk sizes with open_mfdataset

### ISSUE THREAD

**Issue by abishekg7** [pydata/xarray#4445]
I'm using xarray.open_mfdataset to open a dataset of 4 years X 365 days of **ROMS** 3-hourly model output, as follows:
```
files = glob('<prefix>/roms.ocn.hi.201[1-4]*')
ds = xr.open_mfdataset(files, combine='by_coords', compat='override',
                                       concat_dim='ocean_time', data_vars='minimal',
                                       coords='minimal',parallel='True',chunks='auto')
ds.temp
```
And the default chunk sizes are abnormally large. 

![image](https://user-images.githubusercontent.com/56273301/93795358-e2967e80-fbfe-11ea-9a74-326f8a86969e.png)

Since each netcdf file contains 8 time steps, I expected open_mfdataset to form chunks with a max ocean_time:8. I do get the expected behavior when I open a fewer number of files (2 years' worth of data, instead of 4 years).

I also tried specifying the chunk size explicitly with chunks={'ocean_time':8}, but I still have the issue.

![image](https://user-images.githubusercontent.com/56273301/93797682-2dfe5c00-fc02-11ea-8c43-ec9640d7bd3c.png)

**Anything else we need to know?**:
All the model output is in the netCDF-4 format. Disk space: ~2.3G for each day of output, and ~3.2 TB in total. 

<details><summary>Partial output of <tt>ncdump -hs</tt></summary>

dimensions:
	ocean_time = UNLIMITED ; // (8 currently)
	s_rho = 50 ;
	s_w = 51 ;
	tracer = 2 ;
	boundary = 4 ;
	eta_rho = 489 ;
	xi_rho = 655 ;
	eta_u = 489 ;
	xi_u = 654 ;
	eta_v = 488 ;
	xi_v = 655 ;
	eta_psi = 488 ;
	xi_psi = 654 ;

variables:

	double s_rho(s_rho) ;
		s_rho:long_name = "S-coordinate at RHO-points" ;
		s_rho:valid_min = -1. ;
		s_rho:valid_max = 0. ;
		s_rho:positive = "up" ;
		s_rho:standard_name = "ocean_s_coordinate_g2" ;
		s_rho:formula_terms = "s: s_rho C: Cs_r eta: zeta depth: h depth_c: hc" ;
		s_rho:field = "s_rho, scalar" ;
		s_rho:_Storage = "chunked" ;
		s_rho:_ChunkSizes = 50 ;
		s_rho:_DeflateLevel = 1 ;
		s_rho:_Shuffle = "true" ;
		s_rho:_Endianness = "little" ;

	double ocean_time(ocean_time) ;
		ocean_time:long_name = "time since initialization" ;
		ocean_time:units = "seconds since 2010-01-01 00:00:00" ;
		ocean_time:calendar = "gregorian" ;
		ocean_time:field = "time, scalar, series" ;
		ocean_time:_Storage = "chunked" ;
		ocean_time:_ChunkSizes = 512 ;
		ocean_time:_DeflateLevel = 1 ;
		ocean_time:_Shuffle = "true" ;
		ocean_time:_Endianness = "little" ;
       
         float temp(ocean_time, s_rho, eta_rho, xi_rho) ;
		temp:_FillValue = 1.e+37f ;
		temp:long_name = "potential temperature" ;
		temp:units = "Celsius" ;
		temp:time = "ocean_time" ;
		temp:coordinates = "lon_rho lat_rho s_rho ocean_time" ;
		temp:field = "temperature, scalar, series" ;
		temp:_Storage = "chunked" ;
		temp:_ChunkSizes = 1, 17, 163, 219 ;
		temp:_DeflateLevel = 1 ;
		temp:_Shuffle = "true" ;

</details>

**Environment**:

<details><summary>Output of <tt>xr.show_versions()</tt></summary>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.11 | packaged by conda-forge | (default, Aug  5 2020, 20:04:48) 
[GCC 7.5.0]
python-bits: 64
OS: Linux
OS-release: 2.6.32-754.31.1.el6.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None
libhdf5: 1.10.6
libnetcdf: 4.7.4

xarray: 0.16.1
pandas: 1.0.5
numpy: 1.18.5
scipy: 1.5.2
netCDF4: 1.5.3
pydap: None
h5netcdf: None
h5py: 2.10.0
Nio: None
zarr: None
cftime: 1.1.3
nc_time_axis: None
PseudoNetCDF: None
rasterio: None
cfgrib: 0.9.8.4
iris: None
bottleneck: None
dask: 2.25.0
distributed: 2.25.0
matplotlib: 3.2.2
cartopy: 0.18.0
seaborn: None
numbagg: None
pint: None
setuptools: 47.3.1.post20200622
pip: 20.1.1
conda: None
pytest: None
IPython: 7.16.1
sphinx: None
</details>

**Comment by dcherian** [comment]
This  looks like https://github.com/pydata/xarray/issues/4112 but it would only make sense if `temp` wasn't being saved in some later files.

If it works with a subset of files; I would look at the remaining and see what'sgoing wrong.

**Comment by abishekg7** [comment]
Thanks for the suggestion, @dcherian! 

I probed into it a little more, and noticed if I exclude the final month's output, the chunk sizes come out normal

`files = set(glob("<prefix>/roms.ocn.hi.201[1-4]*")) - set(glob("<prefix>/roms.ocn.hi.2014-12*"))`

![image](https://user-images.githubusercontent.com/56273301/93828868-e2fc3d00-fc31-11ea-94d6-934c47261a9f.png)

So far, I haven't noticed anything unusual in the excluded files. No missing fields, etc. And the problem persists across all of the data variables, not just temperature. I'll report back if I find anything.

**Comment by dcherian** [comment]
How about if you just load the last month? What does that look like?

**Comment by abishekg7** [comment]
Just loading the last month with chunks='auto' gives me chunks of size (48, 50, 163, 289) instead of (8, 50, 163, 289). 

![image](https://user-images.githubusercontent.com/56273301/93948314-bd3f6880-fd03-11ea-9186-0cc3a8309b55.png)

**Comment by dcherian** [comment]
that looks like there's one big chunk at the beginning? (see the "rectangle").

Again, this looks like `salt` is missing in these files, and xarray's merge/concat is inserting NaNs to make up for it. This ends up becoming one big chunk with `dask < 2.26.0`

**Comment by abishekg7** [comment]
Ah I see. So far, I've manually gone through each file in that month and `salt` seems to be present in all of them, with reasonable looking values. 

I also noticed that opening the last month of files together with xarray seems to cause the chunking issue, and `ds.salt.isel(ocean_time=3,s_rho=20).plot()` brings up a blank plot. However, xarray seems to get the correct chunk size when opening each netcdf file individually and the plots also look fine. 

Not sure if there's some issue when joining these files.

**Comment by dcherian** [comment]
> ds.salt.isel(ocean_time=3,s_rho=20).plot() brings up a blank plot.

hmm... this is an alignment problem; can you try adding `join="override"`? Is `s_rho` expected to have length 50? Or is that bigger than expected?

My next step would be to read just `salt` from these files and combine them with `xr.concat` and work from there.

**Comment by abishekg7** [comment]
Thank you, @dcherian! 

So I tried `open_mfdataset` with `join="override"`. It works fine with the couple of files in the month that didn't have the issue (12-01 to 12-05). But when I include the file with issue (12-01 to 12-06), I get the following message:

`ValueError: Indexes along dimension 'ocean_time' don't have the same length. Cannot use join='override'.`

I manually confirmed that the `ocean_time` dimension in all six files have the same length (8). And the rest of the dimensions match too. So not sure what's going on here. 

Re. your question, `s_rho` is expected to have a length of 50. I'll try the `xr.concat` approach next. Thanks for your help!
